docs/design/design-discovery.md

```diff
@@ -0,0 +1,100 @@
+
+**Nodes with a firewall and/or NAT device between them (currently in progress):**
+
+```
+                               +---------+
+             +--------+        |Signaling|         +--------+
+             | STUN   |        |Server   |         | STUN   |
+             | Server |        +---------+         | Server |
+             +--------+       /           \        +--------+
+                             /             \
+                            /               \
+                           / <- Signaling -> \
+                          /                   \
+                   +--------+               +--------+
+                   |  NAT   |               |  NAT   |
+                   +--------+               +--------+
+                      /                             \
+                     /                               \
+                 +-------+                       +-------+
+                 | Agent |                       | Agent |
+                 |   L   |                       |   R   |
+                 +-------+                       +-------+
+```
+
+Figure 2 - NAT traversal (we are refering to the signalling server here as a relay node)
+
+- Two endpoints that do not match the initial local address peering, will attempt to peer via the STUN method (see the ICE RFC for further STUN details). A STUN request can be used to open a source UDP port on the NAT device front-ending the endpoint.
+- That UDP source port will remain open for some unknown period of time (depending on the NAT device).
+- Next we attempt to match peer candidates with that IP:Port STUN address for the endpoint.
+- All nodes will attach to a relay node running a STUN server and/or terminating wireguard peerings for fallback forwarding in the case a peer candidate match cannot be made. This peering can be reconciled from wireguard state until we no longer terminate peers on the relay node (tbd).
+- This peering can handle multiple nodes behind a PAT (port address translation) device since there will always be a unique source address assigned in the connection state.
+
+**Nodes that still cannot find a match (currently supported):**
+
+- For all nodes that are unable to find a suitable candidate peering, we fallback to forwarding traffic through the relay node that need to be open (UDP 51820) to all endpoints in the zone.
+- Currently all nodes that peer into a zone, will peer to the relay node to ensure connectivty in all scenarios. Since this is a hairpin, it is the least ideal pairing but can be scaled to fit the small amount of peers that cannot establish a direct peering either through the "Local Address" or STUN discovered "Reflexive Address".
```

Comment:
```suggestion
- Currently, all nodes that peer into a zone will peer to the relay node to ensure connectivity in all scenarios. Since this is a hairpin, it is the least ideal pairing but can be scaled to fit the small number of peers that cannot establish a direct peering either through the "Local Address" or STUN discovered "Reflexive Address".
```